### PR TITLE
checker: extract replica_strategy

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve?
refactor

### What is changed and how it works?
- extract common utilities of `replicaChecker` and `ruleChecker` to `ReplicaStrategy`
- use `ReplicaStrategy` in both checkers
- remove the selector which is not used any more

### Check List

Tests
- Unit test

### Release note
- No release note

### CODE REVIEW

File: server/schedule/checker/replica_strategy.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// SelectStoreToAdd returns the store to add a replica to a region.
+// `coLocationStores` are the stores used to compare location with target
+// store.
+// `extraFilters` is used to set up more filters based on the context that
+// calling this method.
+//
+// For example, to select a target store to replace a region's peer, we can use
+// the peer list without the peer to be removed as `coLocationStores`.
```
Comment (JmPotato) [inline server/schedule/checker/replica_strategy.go:42]: > the peer list without the peer to be removed as `coLocationStores`.

What's the meaning of this sentence?

Comment (disksing) [reply server/schedule/checker/replica_strategy.go:42]: For example, region contains 3 peers `[p1, p2, p3]`. In order to replace `p3`, we will call `SelectStoreToAdd` and pass `[p1, p2]` in.

Comment (JmPotato) [reply server/schedule/checker/replica_strategy.go:42]: I see...So this sentence like this maybe better?

> the peer list with the peer removed as `coLocationStores`.

Comment (disksing) [reply server/schedule/checker/replica_strategy.go:42]: good idea. thanks.

**Pull request merged**